### PR TITLE
Feature/reworked send command #2033

### LINKED ISSUE #2031: Transaction is sent right after tapping on send button (no Send Transaction screen shown)
### Description

*Type*: Bug

*Summary*: No Send transaction screen is shown if send transaction in 1-1 chat. Right after tap on send button there is screen Transaction Sent shown. This issue happens only if previously user sent a transaction from new Wallet.  No issue happens and expected Send Transaction screen is shown if send feature in the new Wallet was not used.

#### Expected behavior
For send command with valid parameters in 1-1 chat: After tap on send button there is screen Send Transaction shown with pre filled data (recipient and amount)
#### Actual behavior
User sees Transaction Sent screen after tapping on send button.

### Reproduction
from 00:26 https://app.testfairy.com/projects/4803590-status/builds/6748397/sessions/3/?accessToken=GoxijWt4cjjsCaKT23q7jfY-eIU
- Open Status
- Open new Wallet
- send some ether to Jarrad from new Wallet
- open 1-1 chat with Jarrad
- tap on send command, select Jarrad as recipient, type valid amount and tap on send button

### Additional Information
* Status version: 
- develop build 0.9.11d171 and 
- PR branch `feature/sign-transaction-in-chat-and-remove-old-wallet` 0.9.11d168
* Operating System: 
Real device iPhone 7 plus, iOs 11.0.2
Real device Google Pixel XL, Android 8.0

### LINKED ISSUE #2033: Wallet - rework send command
image

### PR DESCRIPTION
**Pull request by janherich**
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)
Partially implements #2033 and fixes #2031 

### Summary:
[comment]: # (Summarise the problem and how the pull request solves it)
This PR fixes the most urgent things for the `/send` command to integrate seamlessly into the new wallet.
I started implementing it as a complete move of the sending logic to the new wallet, where `/send` command itself would not be doing the actual adding of the transaction to the tx-queue on the ethereum-node, but just provide necessary arguments for the wallet screen/events to do so. 
All plumbing for that is ready on the command handling side, unfortunately, wallet (screens and transitions between them) is currently coded in such a way, that it expects the present `/send` code behaviour, eq the send modal window is only displayed when transaction is already in tx-queue, therefore I decided to retain that behaviour because of the time constraints. 
I still think we should revisit it after devcon, as the "dumb" `/send` command would make it possible to simplify the wallet code a lot, not even talking about the actual bot js code where the command is defined. 

So what's done:
* When `/send` is issued from the chat and you are taken to the send transaction modal window, upon clicking on "x" icon in the upper left corner, transaction is discarded.
* Fetching the command preview is moved to the later point where command-handler result is ready (tx hash in the case of `/send`), so it provides important prerequisite for implementation of the number of confirmations in the preview + link to the transaction detail from preview
* It removes some obsolete code which is creating by duplicating `/send` command messages for the old wallet. 

What's left to be done:
* Better `/send` command message preview with number of tx confirmations and link to transaction detail

### Steps to test:
- exercise `/send` command behaviour 

[comment]: # (PRs will only be accepted if squashed into single commit.)
status: WIP

### CODE REVIEW

File: resources/js/bots/wallet/bot.js [hunk 1]
```diff
@@ -403,25 +402,29 @@ function handleSend(params, context) {
...
+                },
                 origParams: context["orig-params"]
             });
         } 
     });
-    // async handler, so we don't return anything immediately
+    
+    // jail async handler, we have to signal it
+    return {
+        flow: "jailAsync"
```
Comment (rasom) [inline resources/js/bots/wallet/bot.js:418]: what does this mean?

Comment (rasom) [reply resources/js/bots/wallet/bot.js:418]: the same for "appAsync"

File: src/status_im/ui/screens/wallet/send/views.cljs [hunk 2]
```diff
@@ -23,8 +23,7 @@
 (defn toolbar-view [signing?]
   [toolbar/toolbar2 {:style wallet.styles/toolbar}
    [toolbar/nav-button (act/back-white (if signing?
-                                         #(do (re-frame/dispatch [:wallet/discard-transaction])
-                                              (act/default-handler))
+                                         #(re-frame/dispatch [:wallet/discard-transaction-leave])
```
Comment (flexsurfer) [inline src/status_im/ui/screens/wallet/send/views.cljs]: why did you remove `(act/default-handler)` i think it's expected to close screen and go back to wallet

File: src/status_im/ui/screens/wallet/send/events.cljs [hunk 3]
```diff
@@ -118,16 +118,27 @@
...
-                          :signing? false
-                          :transaction-id nil
-                          :wrong-password? false)}
-             (when transaction-id
-               {:discard-transaction transaction-id})))))
+  (fn [cofx _]
+    (discard-transaction cofx)))
+
+(handlers/register-handler-fx
+  :wallet/discard-transaction-leave
```
Comment (flexsurfer) [inline src/status_im/ui/screens/wallet/send/events.cljs]: tbh i don't like the idea to hide navigation into event, it's better if we can see it in view, as was before

Comment (flexsurfer) [reply src/status_im/ui/screens/wallet/send/events.cljs]: and if we want this it should be named like `discard-transaction-and-navigate-back`, but still i preferer to dispatch 2 events from view, its more clear

File: resources/js/bots/wallet/bot.js [hunk 4]
```diff
@@ -403,25 +402,29 @@ function handleSend(params, context) {
 
     web3.eth.sendTransaction(data, function(error, hash) {
         if (error) {
-            status.sendSignal("handler-result", {
-                status: "failed",
+            /*status.sendSignal("handler-result", {
```
Comment (jeluard) [inline resources/js/bots/wallet/bot.js]: Can we remove this comment?

File: resources/js/bots/wallet/bot.js (hunk 1 above)
Comment (janherich) [reply resources/js/bots/wallet/bot.js:418]: This signalises the workflow which should be taken when evaluating result of the command handler. There will be actually 3 different possibilities:
* Normal synchronous handler, where in the app, upon receiving the result, command-message processing immediately resumes, nothing changes there, we just return data.
* Jail asynchronous handler(the case here), where we return map with `flow` property set to `"jailAsync"` to signalise that processing of command-message should stop and resume only upon receiving signal from jail (`"jail.signal"` of type `"handler-result"`)
* App asynchronous handler (the one I originally planned to use but didn't in the end) where besides flow, you provide `hook` and `data` properties in the return map, this will signalise that you want the processing to go async, but signal to resume won't be from jail, but some module from the RN app (like wallet, contacts, camera capture, etc.).
In the future, we should create and describe such application hooks, which anybody can use when writing their custom commands for status, that way, we could get rid of lot of hardcoded functionality tied to particular commands like it's now.

File: resources/js/bots/wallet/bot.js (hunk 4 above)
Comment (janherich) [reply resources/js/bots/wallet/bot.js]: Yes, of course, this can be removed

File: src/status_im/commands/handlers/jail.cljs [hunk 5]
```diff
@@ -83,20 +75,20 @@
...
-  (handlers/side-effect! suggestions-events-handler!))
+ :suggestions-event!
+ (handlers/side-effect! suggestions-events-handler!))
 
 (reg-handler :set-local-storage
-  (handlers/side-effect!
-    (fn [_ [{:keys [data chat-id]}]]
-      (local-storage/set-data {:chat-id chat-id
-                               :data    data}))))
+             (handlers/side-effect!
```
Comment (flexsurfer) [inline src/status_im/commands/handlers/jail.cljs]: wrong indentation

File: src/status_im/commands/handlers/jail.cljs [hunk 6]
```diff
@@ -83,20 +75,20 @@
...
-  [(after (print-error-message! "Error on param suggestions"))]
-  (handlers/side-effect! suggestions-handler!))
+ :suggestions-handler
+ [(after (print-error-message! "Error on param suggestions"))]
+ (handlers/side-effect! suggestions-handler!))
 
 (reg-handler
-  :suggestions-event!
-  (handlers/side-effect! suggestions-events-handler!))
+ :suggestions-event!
```
Comment (flexsurfer) [inline src/status_im/commands/handlers/jail.cljs]: wrong indentation

File: src/status_im/commands/handlers/jail.cljs [hunk 7]
```diff
@@ -83,20 +75,20 @@
...
-  (after (print-error-message! "Error on command handling"))
-  (handlers/side-effect! command-handler!))
+             (after (print-error-message! "Error on command handling"))
+             (handlers/side-effect! command-handler!))
 
 (reg-handler
-  :suggestions-handler
-  [(after (print-error-message! "Error on param suggestions"))]
-  (handlers/side-effect! suggestions-handler!))
+ :suggestions-handler
```
Comment (flexsurfer) [inline src/status_im/commands/handlers/jail.cljs]: wrong indentation

File: src/status_im/commands/handlers/jail.cljs [hunk 8]
```diff
@@ -83,20 +75,20 @@
       (log/debug message params))))
 
 (reg-handler :command-handler!
-  (after (print-error-message! "Error on command handling"))
-  (handlers/side-effect! command-handler!))
+             (after (print-error-message! "Error on command handling"))
```
Comment (flexsurfer) [inline src/status_im/commands/handlers/jail.cljs]: wrong indentation

**Review by flexsurfer — CHANGES_REQUESTED** [review]

**Review by yenda — APPROVED** [review]
I agree with @flexsurfer comments, the rest works for me

File: src/status_im/ui/screens/wallet/send/views.cljs [hunk 9]
```diff
@@ -148,7 +147,7 @@
 
 (defn toolbar-modal []
   [toolbar/toolbar2 {:style wallet.styles/toolbar}
-   [toolbar/nav-button (act/close-white act/default-handler)]
+   [toolbar/nav-button (act/close-white #(re-frame/dispatch [:wallet/discard-transaction-leave]))]
```
Comment (flexsurfer) [inline src/status_im/ui/screens/wallet/send/views.cljs]: we shouldn't discard transaction because this modal screen can be opened from unsigned transactions in wallet

Comment (flexsurfer) [reply src/status_im/ui/screens/wallet/send/views.cljs]: so this can be done only for send command transaction

File: src/status_im/ui/screens/wallet/transactions/events.cljs [hunk 10]
```diff
@@ -49,7 +49,8 @@
               {:dispatch [:navigate-to-modal :wallet-send-transaction-modal {:amount (str (money/wei->ether value))
                                                                              :transaction-id id
                                                                              :to-address to
-                                                                             :to-name to}]}))))
+                                                                             :to-name to
+                                                                             :in-progress? true}]}))))
```
Comment (flexsurfer) [inline src/status_im/ui/screens/wallet/transactions/events.cljs]: `in-progress?` is used for loader

**Review by flexsurfer — APPROVED** [review]

**Review by rasom — CHANGES_REQUESTED** [review]
Let's not "stop" and "resume" sending of command and get rid of "appAsync" and "jailAsync" stuff first.

**Review by rasom — CHANGES_REQUESTED** [review]
_

**Comment by janherich** [comment]
@rasom So you would like to revert to previous way of signalising async handler by having `asyncHandler` property, instead of special return value, correct ?
Regarding stopping/resuming flow, I'm afraid that that implementing it correctly (so the message status in preview is consistent over network) is quite complicated and it's totally not what we agreed upon on design meeting.
Or at least when I understood you correctly that you would like to `/send` work that way, that immediately after tapping on send, message is displayed in the chat and broadcasted over network.

**Comment by rasom** [comment]
> Regarding stopping/resuming flow, I'm afraid that that implementing it correctly (so the message status in preview is consistent over network) is quite complicated and it's totally not what we agreed upon on design meeting.

Sure, we don't need to implement this now. My concern is why we need to "appAsync"/"jailAsync". I doubt that this functionality is necessary for PR and is needed at all. 

So when i look at 
> So what's done:
>* When `/send` is issued from the chat and you are taken to the send transaction modal window, upon clicking on "x" icon in the upper left corner, transaction is discarded.
>* Fetching the command preview is moved to the later point where command-handler result is ready (tx hash in the case of `/send`), so it provides important prerequisite for implementation of the number of confirmations in the preview + link to the transaction detail from preview
>* It removes some obsolete code which is creating by duplicating `/send` command messages for the old wallet. 

it's not clear which of these items requires "appAsync"/"jailAsync" and why

**Comment by rasom** [comment]
Probably related to `jailAsync`, /faucet command appears after response to this command 

img

**Comment by jeluard** [comment]
@janherich Can you rebase on develop?

**Comment by rasom** [comment]
Do we still need this PR?

**Comment by janherich** [comment]
@rasom Yes, we do for sure, there are many important thing in this PR, the most important is reworked command message flow so the previews are being fetched after handler results are ready, which is necessary to implement better `/send` command previews with link to transaction details and more performant and reliable preview caching.
It won't be included in devcon release for sure, but I will rebase it and make it ready to be merged later this week.

**Comment by oskarth** [comment]
@janherich Please rebase on develop

**Comment by janherich** [comment]
As the code in this PR is very outdated, I'm closing it, as it would be easier to re-implement it from scratch.
